### PR TITLE
[Feature] Support ICNet

### PR DESCRIPTION
**Pull request by MengzhangLI**
Support ICNet

- [x] Refactor
- [x] Add benchmark
- [x] Add unittests
- [x] Adding more unittest to coverage files. 
- [x] Uploading models and logs with correct links

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Sources:
https://github.com/liminn/ICNet-pytorch
https://github.com/hszhao/ICNet
https://github.com/lxtGH/Fast_Seg

image

MMSegmentation:

image

File: configs/icnet/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Icnet for real-time semantic segmentation on high-resolution images
```
Comment (Junjun2016) [inline configs/icnet/README.md:1]:
```suggestion
# ICNet for real-time semantic segmentation on high-resolution images
```

**Comment by MengzhangLI** [comment]
TO DO: Adding more unittest to coverage files.

File: configs/icnet/icnet_r101-d8_832x832_160k_cityscapes.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+_base_ = [
```
Comment (Junjun2016) [inline configs/icnet/icnet_r101-d8_832x832_160k_cityscapes.py:1]: Can derive the configs of r101 and r18 from r50.
E.g. this config can derive from `configs/icnet/icnet_r50-d8_832x832_160k_cityscapes.py`

File: mmseg/models/backbones/icnet.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            Default: None.
+    """
+
+    def __init__(self,
+                 in_channels=3,
+                 layer_channels=(512, 2048),
+                 light_branch_middle_channels=32,
+                 psp_out_channels=512,
+                 out_channels=(64, 256, 256),
+                 backbone_cfg=None,
```
Comment (Junjun2016) [inline mmseg/models/backbones/icnet.py:51]: can make backbone_cfg as a position argument without default.

File: mmseg/models/decode_heads/ic_head.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    """Cascade Feature Fusion Unit in ICNet.
+
+    Args:
+        low_channels (int): The number of input channels for
+            low resolution feature map.
+        high_channels (int): The number of input channels for
+            high resolution feature map.
+        out_channels (int): The number of output channels.
+        num_classes (int): Number of classes for logits.
+        conv_cfg (dict|None): Config of conv layers.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/ic_head.py:21]:
```suggestion
        conv_cfg (dict | None): Config of conv layers.
```

File: mmseg/models/decode_heads/ic_head.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        norm_cfg (dict|None): Config of norm layers.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/ic_head.py:22]:
```suggestion
        norm_cfg (dict | None): Config of norm layers.
```

File: mmseg/models/decode_heads/ic_head.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    """
+
+    def __init__(self,
+                 low_channels,
+                 high_channels,
+                 out_channels,
+                 num_classes,
+                 conv_cfg,
+                 norm_cfg,
+                 act_cfg,
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/ic_head.py:39]: Can add default values for these arguments.

File: tests/test_models/test_backbones/test_icnet.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import pytest
+import torch
+
+from mmseg.models.backbones import ICNet
+
```
Comment (Junjun2016) [inline tests/test_models/test_backbones/test_icnet.py:6]: Add unittests for sub-modules in ICNet.

File: tests/test_models/test_heads/test_ic_head.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+from mmseg.models.decode_heads import ICHead
+from .utils import _conv_has_norm, to_cuda
+
+
```
Comment (Junjun2016) [inline tests/test_models/test_heads/test_ic_head.py:7]: Add unittests for sub-modules in ICHead.

File: mmseg/models/backbones/icnet.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            init_cfg = [
+                dict(type='Kaiming', mode='fan_out', layer='Conv2d'),
+                dict(type='Constant', val=1, layer='_BatchNorm'),
+                dict(type='Normal', mean=0.01, layer='Linear')
+            ]
+        super(ICNet, self).__init__(init_cfg=init_cfg)
+        self.align_corners = align_corners
+        self.backbone = build_backbone(backbone_cfg)
+        self.backbone.maxpool = nn.MaxPool2d(
+            kernel_size=3, stride=2, padding=1, ceil_mode=True)
```
Comment (xvjiarui) [inline mmseg/models/backbones/icnet.py:70]: We may add comments here to explain why changing `maxpool`.

Comment (xvjiarui) [reply mmseg/models/backbones/icnet.py:70]: Add `assert hasattr(self.backbone, 'maxpool')

File: mmseg/models/decode_heads/__init__.py [hunk 10]
```diff
@@ -27,9 +27,31 @@
...
+    'PointHead',
+    'APCHead',
+    'DMHead',
+    'LRASPPHead',
+    'SETRUPHead',
+    'SETRMLAHead',
+    'DPTHead',
+    'SETRMLAHead',
+    'SegformerHead',
+    'ISAHead',
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/__init__.py:56]:
```suggestion
    'ISAHead'
```

File: mmseg/models/necks/ic_neck.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    def __init__(self,
+                 in_channels=(64, 256, 256),
+                 in_index=(0, 1, 2),
+                 out_channels=128,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 align_corners=False,
+                 init_cfg=None):
+        super(ICNeck, self).__init__(init_cfg=init_cfg)
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:110]: should check `len`.

File: mmseg/models/necks/ic_neck.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            Default: dict(type='ReLU').
+        align_corners (bool): align_corners argument of F.interpolate.
+            Default: False.
+        init_cfg (dict or list[dict], optional): Initialization config dict.
+            Default: None.
+    """
+
+    def __init__(self,
+                 in_channels=(64, 256, 256),
+                 in_index=(0, 1, 2),
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:103]: `in_index` not used.

File: mmseg/models/necks/ic_neck.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+            self.out_channels,
+            self.in_channels[0],
+            self.out_channels,
+            conv_cfg=self.conv_cfg,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg,
+            align_corners=self.align_corners)
+
+    def forward(self, inputs):
+        x_sub1, x_sub2, x_sub4 = inputs
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:137]: What if `len(inputs) != 3`?

Comment (MengzhangLI) [reply mmseg/models/necks/ic_neck.py:139]: image

Default structure use 3 feature maps, so I added new assertion on Line 136.

File: mmseg/models/necks/ic_neck.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.conv_cfg = conv_cfg
+        self.norm_cfg = norm_cfg
+        self.act_cfg = act_cfg
+        self.align_corners = align_corners
+        self.cff_24 = CascadeFeatureFusion(
+            self.in_channels[2],
+            self.in_channels[1],
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:119]: If we assume `len(self.in_channels) == 3`, we should have `assert`.

File: mmseg/models/necks/ic_neck.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+                                        maps must be 3!'
+
+        x_sub1, x_sub2, x_sub4 = inputs
+        outputs = list()
+        x_cff_24, x_24 = self.cff_24(x_sub4, x_sub2)
+        outputs.append(x_24)
+        x_cff_12, x_12 = self.cff_12(x_cff_24, x_sub1)
+        outputs.append(x_12)
+        outputs.append(x_cff_12)
+        return outputs
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:146]:
```suggestion
        return x_24, x_12, x_cff_12
```

File: mmseg/models/necks/ic_neck.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+            align_corners=self.align_corners)
+
+        self.cff_12 = CascadeFeatureFusion(
+            self.out_channels,
+            self.in_channels[0],
+            self.out_channels,
+            conv_cfg=self.conv_cfg,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg,
+            align_corners=self.align_corners)
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:133]: `cff_24` and `cff_12` here is confusing. 
Could we make `cff` into a module list?

File: mmseg/models/necks/ic_neck.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        x_low = resize(
+            x_low,
+            size=x_high.size()[2:],
+            mode='bilinear',
+            align_corners=self.align_corners)
+        x_low = self.conv_low(x_low)
+        x_high = self.conv_high(x_high)
+        x = x_low + x_high
+        x = F.relu(x, inplace=True)
+        return x, x_low
```
Comment (xvjiarui) [inline mmseg/models/necks/ic_neck.py:73]: image

Comment (xvjiarui) [reply mmseg/models/necks/ic_neck.py:73]: `conv_low` is `Dilated Conv in the middle of the figure. 
But it shouldn't go into aux loss.

Comment (MengzhangLI) [reply mmseg/models/necks/ic_neck.py:73]: I think our implementation of ICNet is indeed different from this figure. `x_low` does not go through `self.conv_low` but `CLASSIFIER CONV 1x1` in this figure for auxiliary loss.

File: mmseg/models/necks/ic_neck.py (hunk 14 above)
Comment (MengzhangLI) [reply mmseg/models/necks/ic_neck.py:119]: Is `Line 108` OK? I think `assert` has been set.

File: mmseg/models/necks/ic_neck.py (hunk 16 above)
Comment (MengzhangLI) [reply mmseg/models/necks/ic_neck.py:133]: `cff_12` means fusion of `1x` and `2x` feature map followed by [last repo](https://github.com/open-mmlab/mmsegmentation/pull/108/files). Of course `cff` could be made into a module list. But I think it is OK because as a real-time segmentation model, number of `cff` does not needs more because that usually means deeper network and more computation which may make its speed lower.

Comment (xvjiarui) [reply mmseg/models/necks/ic_neck.py:133]: ok

**Review by Junjun2016 — APPROVED** [review]

**Review by xvjiarui — APPROVED** [review]

**Comment by Junjun2016** [comment]
Please remove file [tools/benchmark_new.py](url)
